File: packages/go-service-kit/logging/logging.go
package logging

import (
	"encoding/json"
	"log"
	"os"
)

// Logger is a structured JSON logger.
type Logger struct {
	Service string
}

// NewLogger creates a new Logger.
func NewLogger(service string) *Logger {
	return &Logger{Service: service}
}

// Info logs an info message.
func (l *Logger) Info(msg string, fields map[string]interface{}) {
	entry := map[string]interface{}{
		"level":   "INFO",
		"service": l.Service,
		"message": msg,
	}
	for k, v := range fields {
		entry[k] = v
	}
	b, _ := json.Marshal(entry)
	log.Println(string(b))
}

// Error logs an error message.
func (l *Logger) Error(msg string, err error, fields map[string]interface{}) {
	entry := map[string]interface{}{
		"level":   "ERROR",
		"service": l.Service,
		"message": msg,
		"error":   err.Error(),
	}
	for k, v := range fields {
		entry[k] = v
	}
	b, _ := json.Marshal(entry)
	log.Println(string(b))
}

File: packages/observability/src/observability/logging.py
"""Structured logging for SentraAura.

JSON-structured logs with trace_id propagation.
"""
from __future__ import annotations

import json
import logging
import sys
from typing import Any

from observability.tracing import get_current_trace_id


class JSONFormatter(logging.Formatter):
    """Emit log records as JSON lines."""

    def format(self, record: logging.LogRecord) -> str:
        log_obj: dict[str, Any] = {
            "timestamp": self.formatTime(record),
            "level": record.levelname,
            "logger": record.name,
            "message": record.getMessage(),
            "trace_id": get_current_trace_id() or "",
        }
        if hasattr(record, "extra"):
            log_obj.update(record.extra)
        if record.exc_info:
            log_obj["exception"] = self.formatException(record.exc_info)
        return json.dumps(log_obj, default=str)


def configure_logging(
    level: str = "INFO",
    *,
    service_name: str = "sentra-aura",
    json_format: bool = True,
) -> None:
    """Configure root logger for SentraAura services."""
    root = logging.getLogger()
    root.setLevel(getattr(logging, level.upper(), logging.INFO))

    handler = logging.StreamHandler(sys.stdout)
    if json_format:
        handler.setFormatter(JSONFormatter())
    else:
        handler.setFormatter(
            logging.Formatter("%(asctime)s %(levelname)s %(name)s: %(message)s")
        )

    root.handlers = [handler]


def get_logger(name: str) -> logging.Logger:
    """Get a logger with the given name."""
    return logging.getLogger(name)
